### ISSUE TITLE
About Mixed Precision Training 

### ISSUE THREAD

**Issue by vielzeuf** [speechbrain/speechbrain#612]
Hi,
Thank you for sharing this great framework.
I found it very useful for several ASR experiments.
Yet I had trouble to make the LibriSpeech recipes converging when using auto_mix_prec.

I observe that the "fit_batch" function in core.py (line 825) is written as:
```python
        # Managing automatic mixed precision
        if self.auto_mix_prec:
            with torch.cuda.amp.autocast():
                outputs = self.compute_forward(batch, Stage.TRAIN)
                loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
                self.scaler.scale(loss).backward()
                if self.check_gradients(loss):
                    self.scaler.step(self.optimizer)
                self.optimizer.zero_grad()
                self.scaler.update()
```
and I therefore implement the same fit_batch function in the train.py file of the LibriSpeech recipe. It yields a faster training but no convergence.

According to Pytorch documentation for amp+gradient clipping (https://pytorch.org/docs/stable/notes/amp_examples.html) should be written as:
```python
        # Managing automatic mixed precision
        if self.auto_mix_prec:
            self.optimizer.zero_grad()
            with torch.cuda.amp.autocast():
                predictions = self.compute_forward(batch, Stage.TRAIN)
                loss = self.compute_objectives(predictions, batch, Stage.TRAIN)
            self.scaler.scale(loss).backward()
            self.scaler.unscale_(self.optimizer)
            if self.check_gradients(loss):
                self.scaler.step(self.optimizer)
            self.scaler.update()
```
If you observe the same issue and agree with my observation, I may do a pull request.

Thank you again for the framework,

Best,

Valentin

**Comment by mravanelli** [comment]
Hi @vielzeuf,
I guess you are using the CRDNN model, right? According to our experience, using automatic mixed precision with RNNs leads to convergence issues such as the ones you mentioned. That's why it is set to False by default in this case. In other cases, where you only have feedforward connections (see Transformers, xvectors, ECAPA)  the mixed precision works well.  I think this is due to the recurrent connections, that are a very critical part of the model that should stay with full precision.

**Comment by vielzeuf** [comment]
Thank you for your answer.

To be more specific, I mainly try it with a ContextNet encoder trained by CTC+BPE, trying to implement something similar to the Nemo Citrinet (https://github.com/NVIDIA/NeMo/tree/main/examples/asr/conf/citrinet ) and without RNN. I agree about RNN and mixed precision not doing well together. The proposed change allows this model to converge with amp, when it was not possible with the former implementation. Yet I agree that it does not mean that the proposed change is working for other recipes. I will try it on the Transformer recipe and keep you posted.

I think that the main problem of the current implementation is that it differs with what is provided in Pytorch examples.  For instance, it does not take gradient clipping into account (it is needed to apply an unscale step before clipping to get the same behaviour as without amp). 
Why not using the official way to apply amp as shown in Pytorch documentation (https://pytorch.org/docs/stable/notes/amp_examples.html#amp-examples )? Was it not working ? Or am I missing something ?

**Comment by mravanelli** [comment]
Interesting. We used the current AMP implementation successfully in many other recipes (e.g, ECAPA and Transformers) so I think we assumed every was fine. It could be great to double check if everything is fine with the change you propose as well. @pplantinga, could you double check this change vs the current implementation?
Please, keep me updated on the Citrinet part as well. It would be interesting to see the performance that you can achieve with that.

**Comment by mravanelli** [comment]
@vielzeuf, could you please ask for a PR with this change? Before merging it, we have to make sure it doesn't harm the performance that we achieve with the models currently using AMP (ASR transformers, ECAPA, SepFormer,...)
